api/types/authentication.go

```diff
@@ -701,3 +714,122 @@ func (d *MFADevice) MarshalJSON() ([]byte, error) {
+
+// const (
+// 	// RequireMFAOff means MFA is *not* required to begin server sessions.
+// 	RequireMFAOff RequireMFATypeString = "off"
+// 	// RequireMFASession means MFA is required to begin server sessions.
+// 	RequireMFASession RequireMFATypeString = "session_mfa"
+// 	// RequireMFASessionAndHardwareKey means MFA is required to begin server sessions,
+// 	// and login sessions must use a private key backed by a hardware key.
+// 	RequireMFASessionAndHardwareKey RequireMFATypeString = "hardware_key"
+// 	// RequireMFAHardwareKeyTouch means login sessions must use a hardware private key that
+// 	// requires touch to be used. This touch requirement applies to all API requests
+// 	// rather than only session requests. This touch is different from MFA, so to prevent
+// 	// requiring double touch on session requests, normal Session MFA is disabled.
+// 	RequireMFAHardwareKeyTouch RequireMFATypeString = "hardware_key_touch"
+// )
+
+// IsSessionMFARequired returns whether this RequireMFAType requires per-session MFA.
+func (r RequireMFAType) IsSessionMFARequired() bool {
+	return r == RequireMFAType_SESSION || r == RequireMFAType_SESSION_AND_HARDWARE_KEY
+}
+
+// MarshalJSON marshals RequireMFAType to boolean or string.
+func (r *RequireMFAType) MarshalYAML() (interface{}, error) {
+	val, err := r.encode()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return val, nil
+}
+
+// UnmarshalYAML supports parsing RequireMFAType from boolean or alias.
+func (r *RequireMFAType) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var val interface{}
+	err := unmarshal(&val)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = r.decode(val)
+	return trace.Wrap(err)
+}
+
+// MarshalJSON marshals RequireMFAType to boolean or string.
+func (r *RequireMFAType) MarshalJSON() ([]byte, error) {
+	val, err := r.encode()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return json.Marshal(val)
+}
+
+// UnmarshalJSON supports parsing RequireMFAType from boolean or alias.
+func (r *RequireMFAType) UnmarshalJSON(data []byte) error {
+	var val interface{}
+	err := json.Unmarshal(data, &val)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = r.decode(val)
+	return trace.Wrap(err)
+}
+
+var (
+	RequireMFATypeHardwareKeyString      = "hardware_key"
+	RequireMFATypeHardwareKeyTouchString = "hardware_key_touch"
+)
+
+// encode RequireMFAType into a string or boolean. This is necessary for
+// backwards compatibility with the json/yaml tag "require_session_mfa",
+// which used to be a boolean.
+func (r *RequireMFAType) encode() (interface{}, error) {
+	switch *r {
+	case RequireMFAType_OFF:
+		return false, nil
+	case RequireMFAType_SESSION:
+		return true, nil
+	case RequireMFAType_SESSION_AND_HARDWARE_KEY:
+		return RequireMFATypeHardwareKeyString, nil
+	case RequireMFAType_HARDWARE_KEY_TOUCH:
+		return RequireMFATypeHardwareKeyTouchString, nil
+	default:
+		return nil, trace.BadParameter("RequireMFAType invalid value %v", r)
```

Comment:
```suggestion
		return nil, trace.BadParameter("RequireMFAType invalid value %v", *r)
```
deref pointer in error message